Fix setting timeout in ansible.cfg ssh_connection section

s-hertel opened the pull request:
##### SUMMARY
 Attempt to fix a couple issues that prevent setting `timeout` in the `ssh_connection` ansible.cfg section (though, this could apply to any connection plugin).
 
* The CLI option `--timeout` overrides other configuration methods.
  Since `--timeout` had a default, PC would always use this value.

* FAs preferred parent FA defaults even if they defined their own.
  PlayContext uses the same name as a FA for Base, but they're unrelated (connection timeout vs task timeout), so PC should be able to set its own attribute options.

Related to #76590 (which fixed magic vars taking precedence too).

Connection plugin options that use the key `timeout` also couldn't set their own default since PlayContext.timeout was always passed in task_keys.

Still looking if there's a way to fix the value the config lookup retrieves.

##### ISSUE TYPE
- Bugfix Pull Request

s-hertel committed:
WIP - fix setting timeout in ansible.cfg ssh_connection section

bcoca commented:
the connection should define 'cli' and 'timeout' to get the proper values  (as well as ini/env) , this is fine as a workaround but should issue a deprecation (connections should update to use correct form)

bcoca commented on lib/ansible/executor/task_executor.py:
```diff
@@ -1051,22 +1051,23 @@ def _set_connection_options(self, variables, templar):
 
         # set options with 'templated vars' specific to this plugin and dependent ones
         self._connection.set_options(task_keys=task_keys, var_options=options)
+        if 'timeout' in self._connection._options and self._connection.get_option('timeout') is None:
```
dont inspect _options, get the plugin config defs and see if 'timeout' is one of the options (also allows you to check for cli/env/ini entries)

s-hertel replied:
I wonder if it should just be a warning/error for a plugin to have anything like:
```
cli:
  - name: timeout  # defaults to DEFAULT_TIMEOUT, which defaults to 10
default: anything that isn't 10
```
This PR is special-casing --timeout, but it would be the same for any cli option that has a default value.

Also wondering what should happen if a plugin has multiple items in an option's ``cli`` list? If one of those has a default value, it won't ever consider the rest in the list currently. Maybe that should be a warning/error too.

bcoca replied:
`--user` and others already exist, the problem is mainly that the default defined in the global config was different than the plugin one (also the conflict with the `timeout` keyword .. that was my bad).

multiple `cli` entries would use the same precedence as other entries .. not that i can see a world in which that makes sense. the bigg diff with these and `env` and the others is that they are not 'autocreated' for the user, they must be preexisting in the argsparser dev.

s-hertel replied:
The ssh plugin timeout option has the correct default (same as the global default), I'm just trying to understand the big picture since it's the same cause. It does seem like the plugin should just document the global default and otherwise it's wrong, but I don't know how to reconcile that with also being unable to set an option using any method that would happen following https://github.com/ansible/ansible/blob/devel/lib/ansible/config/manager.py#L498-L502 if the cli option has a default. The option is unsettable via env, ini, default (only settable with direct, vars, keyword). So maybe it's not a bug timeout can't be set in [ssh_connection], maybe it's just a docs issue for `ssh` and we need some kind of safeguard in validate-modules to prevent misunderstanding.

And multiple `cli` options is far-fetched, just wanted an idea of how it was supposed to work since it is a list. :)

bcoca replied:
i would move from using the default in the cli options and allowing the plugins to diverge, eventually removing the 'global options'

s-hertel committed:
Check the plugin options that contain the cli option timeout instead of assuming the option name

Don't abuse task_keys to set defaults for non-pb keywords

Leave a comment that this seems wrong - should be fixed generically

s-hertel committed:
180, remove timeout-specific hack

s-hertel commented on lib/ansible/cli/arguments/option_helpers.py:
```diff
@@ -252,7 +252,7 @@ def add_connect_options(parser):
     connect_group.add_argument('-c', '--connection', dest='connection', default=C.DEFAULT_TRANSPORT,
                                help="connection type to use (default=%s)" % C.DEFAULT_TRANSPORT)
-    connect_group.add_argument('-T', '--timeout', default=C.DEFAULT_TIMEOUT, type=int, dest='timeout',
+    connect_group.add_argument('-T', '--timeout', type=int, default=C.DEFAULT_TIMEOUT, dest='timeout',
```
Fixing diff
```suggestion
    connect_group.add_argument('-T', '--timeout', default=C.DEFAULT_TIMEOUT, type=int, dest='timeout',
```

s-hertel committed:
Update lib/ansible/cli/arguments/option_helpers.py

s-hertel committed:
WIP - fix cli defaults overriding plugin-specific defaults
